rfc-116-store-attachment-data-in-content-items.md

```diff
@@ -0,0 +1,391 @@
+   1. Add the field as optional.
+   2. Update publishing apps (just Whitehall and Content Publisher) to write to it.
+   3. Republish all affected documents.
+   4. Make the field mandatory.
+
+1. Remove the `details.documents` field.
+
+   **Migration concerns:** this would need to be done in four steps:
+   1. Update frontend apps to render from
+      `details.featured_attachments`.
+   2. Update publishing apps to not set `details.documents`.
+   3. Republish all affected documents.
+   4. Remove the field.
+
+The final schemas will be:
+
+```
+{
+  image_asset: {
+    type: "object",
+    additionalProperties: false,
+    required: [
+      "content_type",
+      "url",
+    ],
+    properties: {
+      alt_text: { type: "string", },
+      caption: { type: "string", },
+      content_type: { type: "string", },
+      credit: { type: "string", },
+      url: { type: "string", format: "uri", },
+    },
+  },
+
+  file_attachment_asset: {
+    type: "object",
+    additionalProperties: false,
+    required: [
+      "attachment_type",
+      "content_type",
+      "url",
+    ],
+    properties: {
+      accessible: { type: "boolean", },
+      alternative_format_contact_email: { type: "string", },
+      attachment_type: { type: "string", enum: ["file"], },
+      content_type: { type: "string", },
+      file_size: { type: "integer", },
+      filename: { type: "string", },
+      id: { type: "string" },
+      locale: { "$ref": "#/definitions/locale", },
+      number_of_pages: { type: "integer", },
+      preview_url: { type: "string", format: "uri", },
+      title: { type: "string", },
+      url: { type: "string", format: "uri", },
+    },
+  },
+
+  html_attachment_asset: {
+    type: "object",
+    additionalProperties: false,
+    required: [
+      "attachment_type",
+      "url",
+    ],
+    properties: {
+      attachment_type: { type: "string", enum: ["html"], },
+      id: { type: "string" },
+      locale: { "$ref": "#/definitions/locale", },
+      title: { type: "string", },
+      url: { type: "string", format: "uri", },
+    },
+  },
+
+  external_attachment_asset: {
+    type: "object",
+    additionalProperties: false,
+    required: [
+      "attachment_type",
+      "url",
+    ],
+    properties: {
+      attachment_type: { type: "string", enum: ["external"], },
+      id: { type: "string" },
+      locale: { "$ref": "#/definitions/locale", },
+      title: { type: "string", },
+      url: { type: "string", format: "uri", },
+    },
+  },
+
+  specialist_publisher_attachment_asset: {
+    allOf: [
+      { "$ref": "#/definitions/file_attachment_asset" },
+      {
+        required: [
+          "content_id",
+        ],
+        properties: {
+          content_id: { "$ref": "#/definitions/guid", },
+          created_at: { format: "date-time", },
+          updated_at: { format: "date-time", },
+        },
+      },
+    ],
+  },
+
+  publication_attachment_asset: {
+    allOf: [
```

Comment:
I don't think this allOf is going to quite work because the various attachment asset schemas are set to have `additionalProperties: false` which would mean they fail the oneOf schema if they have any of the publication properties. 

I am impressed though with this allOf approach, I didn't know about this technique to allow inheritance - very clever.

Reply:
Hmm, that's annoying.  The json-schema maintainer says that json-schema isn't for inheritance (https://github.com/json-schema-org/json-schema-spec/issues/348#issuecomment-322940347), so there might be no better solution than duplicating all the fields.

Reply:
Sure - field duplication is how we've solved this problem in the past so would be consistent with our prior approaches.

As we use [jsonnet](https://jsonnet.org/) to define the schema definitions we can cut this duplication out of the code that manages schemas. For instance, my initial expectation of this base_attachment_asset was that it'd be something that existed in jsonnet.

Reply:
:+1: